concept/tests/test_thing.rs

```diff
@@ -655,3 +741,283 @@ fn role_player_duplicates() {
+            _ => assert!(false, "Wrong data type"),
+        }
+
+        let attr_by_id = thing_manager
+            .get_attribute_with_value(&snapshot, attr_type.clone(), Value::Struct(Cow::Borrowed(&struct_value)))
+            .unwrap()
+            .unwrap();
+        assert_eq!(attr, attr_by_id);
+        assert_eq!(attr_created, attr);
+        snapshot.close_resources();
+    }
+}
+
+#[test]
+fn read_struct_by_field() {
+    let (storage, type_manager, thing_manager) = prepare();
+
+    let attr_label = Label::build("index_test_attr");
+    let nested_struct_spec =
+        ("nested_test_struct".to_owned(), HashMap::from([("nested_string".to_owned(), (ValueType::String, false))]));
+
+    let nested_struct_key = {
+        let mut snapshot = storage.clone().open_snapshot_write();
+        let nested_struct_key = define_struct(&mut snapshot, &type_manager, nested_struct_spec.0, nested_struct_spec.1);
+        snapshot.commit().unwrap();
+        nested_struct_key
+    };
+
+    let struct_spec = (
+        "index_test_struct".to_owned(),
+        HashMap::from([("f_nested".to_owned(), (ValueType::Struct(nested_struct_key.clone()), false))]),
+    );
+    let (attr_type, struct_key, struct_def) = {
+        let mut snapshot = storage.clone().open_snapshot_write();
+        let struct_key = define_struct(&mut snapshot, &type_manager, struct_spec.0, struct_spec.1);
+        let mut attr_type = type_manager.create_attribute_type(&mut snapshot, &attr_label, false).unwrap();
+        attr_type.set_value_type(&mut snapshot, &type_manager, ValueType::Struct(struct_key.clone())).unwrap();
+        attr_type
+            .set_annotation(&mut snapshot, &type_manager, AttributeTypeAnnotation::Independent(AnnotationIndependent))
+            .unwrap();
+        let struct_def = type_manager.get_struct_definition(&snapshot, struct_key.clone()).unwrap();
+        snapshot.commit().unwrap();
+        (attr_type, struct_key.clone(), struct_def)
+    };
+
+    // Create value
+    let field_values = ["abc", "xyz_but_long_enough_to_hash"];
+    {
+        let mut snapshot = storage.clone().open_snapshot_write();
+        let mut attrs = Vec::new();
+        for val in field_values {
+            let nested_struct_value = StructValue::new(
+                nested_struct_key.clone(),
+                HashMap::from([(0, FieldValue::String(Cow::Borrowed(val)))]),
+            );
+            let outer_struct_value = StructValue::new(
+                struct_key.clone(),
+                HashMap::from([(0, FieldValue::Struct(Cow::Owned(nested_struct_value)))]),
+            );
+            let attr = thing_manager
+                .create_attribute(&mut snapshot, attr_type.clone(), Value::Struct(Cow::Borrowed(&outer_struct_value)))
+                .unwrap();
+            attrs.push(attr);
+        }
+
+        for (val, attr) in std::iter::zip(field_values, attrs.iter()) {
+            let field_path = type_manager
+                .resolve_struct_field(
```

Comment:
I think this should not take a vec of owned slices, it should take a vec of slices (or even a slice of slices?) - conceptually, we should not need to take ownership to produce the output data type.